enterprise_access/apps/subsidy_access_policy/models.py

```diff
@@ -1048,7 +1058,44 @@ def can_redeem(self, lms_user_id, content_key, skip_customer_user_check=False):
         return (True, None, existing_redemptions)
 
     def redeem(self, lms_user_id, content_key, all_transactions, metadata=None):
-        raise NotImplementedError
+        """
+        Redeem content, but only if there's a matching assignment.  On successful redemption, the assignment state will
+        be set to 'accepted', otherwise 'errored'.
+
+        Returns:
+            A ledger transaction.
+
+        Raises:
+            SubsidyAPIHTTPError if the Subsidy API request failed.
+            ValueError if the access method of this policy is invalid.
+        """
+        found_assignment = assignments_api.get_assignment_for_learner(
+            self.assignment_configuration,
+            lms_user_id,
+            content_key,
+        )
+        # The following checks for non-allocated assignments only exist to be defensive against race-conditions, but
+        # in practice should never happen if the caller locks the policy and runs can_redeem() before redeem().
+        if not found_assignment:
+            raise MissingAssignment(
+                f'No assignment was found for lms_user_id={lms_user_id} and content_key=<{content_key}>.'
+            )
+        if found_assignment.state != LearnerContentAssignmentStateChoices.ALLOCATED:
+            raise MissingAssignment(
+                f"Only an assignment with state='{found_assignment.state}' was found for lms_user_id={lms_user_id} "
+                f"and content_key=<{content_key}>."
+            )
+        try:
+            ledger_transaction = super().redeem(lms_user_id, content_key, all_transactions, metadata)
+            # Migrate assignment to accepted.
+            found_assignment.state = LearnerContentAssignmentStateChoices.ACCEPTED
```

Comment:
We should also record the transaction uuid on `found_assignment`.

Reply:
ah, forgot about that